### PR TITLE
Multipart parser

### PR DESCRIPTION
**Pull request by sunharis**
As part of HMC performance improvement, this commit implements
multiple configfile upload with a single http request

This commit uses the upstream MIME parsing algorithm.

Upstream commit: https://gerrit.openbmc-project.xyz/c/openbmc/bmcweb/+/34974
The 2nd commit is a downstream only based on https://gerrit.openbmc-project.xyz/c/openbmc/bmcweb/+/32174/

Tested by:
    1. Multiple file upload
    curl -k -H "X-Auth-Token: $bmc_token" -H "Content-Type: multipart/form-data"
    -X POST https://${bmc}/ibm/v1/Host/ConfigFiles -F "testfile1=@<path>;type=application/octet-stream"
    -F "testfile2=@<path>;type=application/octet-stream"
    2. Tested the single file upload for sanity

### CODE REVIEW

**Comment by sunharis** [comment]
This change could potentially fix the performance issue https://w3.rchland.ibm.com/projects/bestquest/?defect=SW538927

**Comment by gtmills** [comment]
I updated the PR msg to add links. For next time, please do that. 
My concern with  https://gerrit.openbmc-project.xyz/c/openbmc/bmcweb/+/34974, is it still has unaddressed valid comments. 
I don't see any of the 7 that are a show stopper but they would help maintainability. Could these be addressed? @lxwinspur

File: include/ibm/management_console_rest.hpp [hunk 1]
```diff
@@ -1,9 +1,10 @@
 #pragma once
```
Comment (gtmills) [inline include/ibm/management_console_rest.hpp:1]: Could you add some more detail to the commit msg about exactly what you are doing here? 
Where it is from, etc.

File: include/ibm/management_console_rest.hpp [hunk 2]
```diff
@@ -1075,6 +1160,14 @@ inline void requestRoutes(App& app)
                 handleFileUrl(req, asyncResp, fileName);
             });
 
+    BMCWEB_ROUTE(app, "/ibm/v1/Host/ConfigFiles")
```
Comment (gtmills) [inline include/ibm/management_console_rest.hpp:1163]: Why before were we posting to /ibm/v1/Host/ConfigFiles/<str> ?

Comment (gtmills) [reply include/ibm/management_console_rest.hpp:1163]: So Code update is doing multipart at 

    "MultipartHttpPushUri": "/redfish/v1/UpdateService/upload",
    
    Does it make sense for us to follow?

Comment (gtmills) [reply include/ibm/management_console_rest.hpp:1163]: Should we advertise this in the GET somehow like Code update?

File: include/ibm/management_console_rest.hpp [hunk 3]
```diff
@@ -238,14 +172,165 @@ inline void handleFilePut(const crow::Request& req,
...
+
+    std::filesystem::path loc(configFilePath);
+
+    // Get the current size of the savearea directory
+    std::filesystem::recursive_directory_iterator iter(loc, ec);
+    if (ec)
+    {
+        asyncResp->res.result(
+            boost::beast::http::status::internal_server_error);
+        asyncResp->res.jsonValue["Description"] = internalServerError;
```
Comment (gtmills) [inline include/ibm/management_console_rest.hpp:219]: We could add more here.. These Descriptions are all the same. Up to you

File: include/ibm/management_console_rest.hpp [hunk 4]
```diff
@@ -238,14 +172,165 @@ inline void handleFilePut(const crow::Request& req,
...
+        std::string fileName;
+        for (const FormPart& formpart : parser.mime_fields)
+        {
+            boost::beast::http::fields::const_iterator it =
+                formpart.fields.find("Content-Disposition");
+            if (it == formpart.fields.end())
+            {
+                BMCWEB_LOG_ERROR << "Couldn't find Content-Disposition";
+                asyncResp->res.result(
+                    boost::beast::http::status::internal_server_error);
```
Comment (gtmills) [inline include/ibm/management_console_rest.hpp:281]: Sure seems like not an internal error and instead a bad request

File: include/ibm/management_console_rest.hpp [hunk 5]
```diff
@@ -238,14 +172,165 @@ inline void handleFilePut(const crow::Request& req,
...
+    {
+        BMCWEB_LOG_DEBUG << "This is a multipart upload";
+        MultipartParser parser(req, ec);
+        if (ec)
+        {
+            // handle error
+            BMCWEB_LOG_ERROR << "MIME parse failed, ec : " << ec.value()
+                             << " , ec message : " << ec.message();
+            asyncResp->res.result(
+                boost::beast::http::status::internal_server_error);
```
Comment (gtmills) [inline include/ibm/management_console_rest.hpp:267]: Really an internal error and not a bad request?

**Comment by gtmills** [comment]
Comments upstream, returning internal error at times when I don't think we should but overall I am good with the direction

File: include/ibm/management_console_rest.hpp (hunk 2 above)
Comment (sunharis) [reply include/ibm/management_console_rest.hpp:1163]: There is already a GET and PATCH  command implemented on this URL for the save area file upload and download purpose. This is a new POST command to upload multiple files at one shot. 
Code update is the redfish path. This /ibm/v1 is a REST way of implementing the routes.

**Comment by lxwinspur** [comment]
> I updated the PR msg to add links. For next time, please do that. My concern with https://gerrit.openbmc-project.xyz/c/openbmc/bmcweb/+/34974, is it still has unaddressed valid comments. I don't see any of the 7 that are a show stopper but they would help maintainability. Could these be addressed? @lxwinspur

@gtmills Okay

**Comment by lxwinspur** [comment]
> Comments upstream, returning internal error at times when I don't think we should but overall I am good with the direction

Let me update upstream

File: include/ibm/management_console_rest.hpp (hunk 2 above)
Comment (gtmills) [reply include/ibm/management_console_rest.hpp:1165]: I would prefer we match code update here...
"MultipartHttpPushUri": "/redfish/v1/UpdateService/upload",

File: include/ut/multipart_test.cpp [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+                 "-----------------------------d74496d66958873e\r\n"
+                 "Content-Disposition: form-data; name=\"Test2\"\r\n"
+                 "123456\r\n"
+                 "-----------------------------d74496d66958873e--\r\n";
+
+    std::error_code ec;
+    crow::Request reqIn(req);
+    MultipartParser parser(reqIn, ec);
+    ASSERT_EQ(ec.value(), static_cast<int>(ParserError::ERROR_EMPTY_HEADER));
+}
```
Comment (gtmills) [inline include/ut/multipart_test.cpp:62]: There is comment upstream about more tests, I think that is valid

File: include/ut/multipart_test.cpp [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    ASSERT_EQ(parser.mime_fields[0].fields.at("Content-Disposition"),
+              "form-data; name=\"Test1\"");
+    ASSERT_EQ(parser.mime_fields[0].content, "{\"Key1\": 112233}");
+
+    ASSERT_EQ(parser.mime_fields[1].fields.at("Content-Disposition"),
+              "form-data; name=\"Test2\"");
+    ASSERT_EQ(parser.mime_fields[1].content, "123456");
+}
+
+TEST(MultipartTest, TestBadMultipartParser)
```
Comment (gtmills) [inline include/ut/multipart_test.cpp:42]: Valid comment here about the name

**Pull request merged**